crates/client/eth/src/l1_messaging.rs

```diff
@@ -0,0 +1,497 @@
+use anyhow::Context;
+use futures::StreamExt;
+use std::sync::Arc;
+
+use crate::client::StarknetCoreContract::LogMessageToL2;
+use crate::client::{EthereumClient, StarknetCoreContract};
+use crate::utils::u256_to_felt;
+use alloy::primitives::{keccak256, FixedBytes, U256};
+use alloy::sol_types::SolValue;
+use blockifier::transaction::transactions::L1HandlerTransaction as BlockifierL1HandlerTransaction;
+use dc_db::{l1_db::LastSyncedEventBlock, DeoxysBackend};
+use dp_utils::channel_wait_or_graceful_shutdown;
+use starknet_api::core::{ChainId, ContractAddress, EntryPointSelector, Nonce};
+use starknet_api::transaction::{
+    Calldata, Fee, L1HandlerTransaction, Transaction, TransactionHash, TransactionVersion,
+};
+use starknet_api::transaction_hash::get_transaction_hash;
+use starknet_types_core::felt::Felt;
+
+pub async fn sync(backend: &DeoxysBackend, client: &EthereumClient, chain_id: &ChainId) -> anyhow::Result<()> {
+    tracing::info!("⟠ Starting L1 Messages Syncing...");
+
+    let last_synced_event_block = match backend.messaging_last_synced_l1_block_with_event() {
+        Ok(Some(blk)) => blk,
+        Ok(None) => {
+            unreachable!("Should never be None")
+        }
+        Err(e) => {
+            tracing::error!("⟠ Madara Messaging DB unavailable: {:?}", e);
+            return Err(e.into());
+        }
+    };
+    let event_filter = client.l1_core_contract.event_filter::<StarknetCoreContract::LogMessageToL2>();
+    let mut event_stream = event_filter
+        .from_block(last_synced_event_block.block_number)
+        .watch()
+        .await
+        .context("Failed to watch event filter")?
+        .into_stream();
+
+    while let Some(event_result) = channel_wait_or_graceful_shutdown(event_stream.next()).await {
+        if let Ok((event, meta)) = event_result {
+            tracing::info!(
+                "⟠ Processing L1 Message from block: {:?}, transaction_hash: {:?}, log_index: {:?}, fromAddress: {:?}",
+                meta.block_number,
+                meta.transaction_hash,
+                meta.log_index,
+                event.fromAddress
+            );
+
+            // Check if cancellation was initiated
+            let event_hash = get_l1_to_l2_msg_hash(&event)?;
+            tracing::info!("⟠ Checking for cancelation, event hash : {:?}", event_hash);
+            let cancellation_timestamp = client.get_l1_to_l2_message_cancellations(event_hash).await?;
+            if cancellation_timestamp != Felt::ZERO {
+                tracing::info!("⟠ L1 Message was cancelled in block at timestamp : {:?}", cancellation_timestamp);
```

Comment:
not sure to understand here, do you mean if we get two time the same event before the process end ?
because nonce is checked right after at the very beginning of process_l1_message and to me, last synced block should be push to db once all the processing went smooth but maybe i'm wrong. What do you think ?

Reply:
So my view was that if we have a lot of cancelled messages on the L1 then our code will read the messages but do nothing. It might be a good idea to store in DB that we've now processed these messages (which is do nothing in the case of cancelled messages) so that we don't run them again if we restart

Reply:
done

Reply:
should we set last synced also

Reply:
is it really "synced" if it has been cancelled ? how do we differentiate message that are cancelled or not ?